src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Query.cs

```diff
@@ -316,16 +401,29 @@ public Task<ResultSetSubset> GetSubset(int batchIndex, int resultSetIndex, int s
 
                 try
                 {
+                    // Execute beforeBatches synchronously, before the user defined batches 
+                    foreach (Batch b in BeforeBatches)
+                    {
+                        await b.Execute(conn, cancellationSource.Token);
+                    }
+
                     // We need these to execute synchronously, otherwise the user will be very unhappy
                     foreach (Batch b in Batches)
                     {
+                        // Attach extension callbacks to the these batches because the user explicity ran them
```

Comment:
I'll rephrase. I was just trying to say that these have completion handlers because they aren't injected batches like the beforeBatches and afterBatches.